graspologic/cluster/rclust.py

```diff
@@ -0,0 +1,320 @@
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+        level: int, optional (default=None)
+            the level of flat clustering to generate
+
+        Returns
+        -------
+        labels : array-like, shape (n_samples, n_levels)
+            if no level specified; otherwise, shape (n_samples,)
+        """
+
+        check_is_fitted(self, ["model_"], all_or_any=all)
+
+        X = check_array(X, dtype=[np.float64, np.float32], ensure_min_samples=1)
+        if level is not None:
+            if not isinstance(level, int):
+                raise TypeError("level must be an int")
+            elif level < 1:
+                raise ValueError("level must be positive")
+
+        labels = self._predict_labels(X)
+        labels = self._relabel(labels)
+
+        if level is not None:
+            if level <= labels.shape[1]:
+                labels = labels[:, level - 1]
+            else:
+                msg = "input exceeds max level = {}".format(labels.shape[1])
+                raise ValueError(msg)
+
+        return labels
+
+    def _predict_labels(self, X):
+        if not self.is_leaf:
+            pred_labels = np.zeros((len(X), self.height), dtype=int)
+            current_pred_labels = self.model_.predict(X)
+            pred_labels[:, 0] = current_pred_labels
+            for label in np.unique(current_pred_labels):
+                current_child = self.children[label]
+                if not current_child.is_leaf:
+                    child_pred_labels = current_child._predict_labels(
+                        X[current_pred_labels == label]
+                    )
+                    pred_labels[
+                        current_pred_labels == label, 1 : child_pred_labels.shape[1] + 1
+                    ] = child_pred_labels
+        else:
+            # only for cases where root is a leaf cluster, i.e.,
+            # only 1 cluster predicted at 1st level
+            if self.is_root:
+                pred_labels = np.zeros((len(X), 1), dtype=int)
+
+        return pred_labels
+
+    def _relabel(self, labels):
+        # re-number labeling so that labels[:,i] assigns a unique value to each cluster
+        n_levels = labels.shape[1]
+        for level in range(n_levels):
+            # to avoid updated labels accidentally matching old ones
+            new_labels = -labels[:, : level + 1] - 1
+            uni_labels = np.unique(new_labels, axis=0)
+            for ul in range(len(uni_labels)):
+                uni_labels_inds = np.where(
+                    (uni_labels[:, None, :] == new_labels).all(2)[ul]
+                )[0]
+                labels[uni_labels_inds, level] = ul
+
+        # delete labels[:,-1] b/c it will be the same as labels[:,-2]
+        # if n_components=1 for pred of every cluster at level of depth=1
+        if labels.shape[1] > 1:
+            if np.max(labels[:, -1]) == np.max(labels[:, -2]):
+                labels = labels[:, :-1]
```

Comment:
@bdpedigo this part is unclear to me, but i dont have a clear idea of what to suggest, reminder to revisit